restorer_test.go

```diff
@@ -189,89 +212,208 @@ func testRestorer(t *testing.T, when spec.G, it spec.S) {
-				h.AssertNil(t, restorer.Restore(testCache))
-				if _, err := os.Stat(filepath.Join(layersDir, "nogroup.buildpack.id")); !os.IsNotExist(err) {
-					t.Fatal("Error: expected nogroup.buildpack.id not to be restored")
-				}
+			when("there is a cache=true layer not in cache", func() {
+				it.Before(func() {
+					meta := "cache=true"
+					h.AssertNil(t, writeLayer(layersDir, "buildpack.id", "cache-layer-not-in-cache", meta, "some-made-up-sha"))
+					h.AssertNil(t, restorer.Restore(testCache))
+				})
+
+				it("removes metadata and sha file", func() {
+					h.AssertPathDoesNotExist(t, filepath.Join(layersDir, "buildpack.id", "cache-layer-not-in-cache.toml"))
+					h.AssertPathDoesNotExist(t, filepath.Join(layersDir, "buildpack.id", "cache-layer-not-in-cache.sha"))
+				})
+				it("does not restore layer data", func() {
+					h.AssertPathDoesNotExist(t, filepath.Join(layersDir, "buildpack.id", "cache-layer-not-in-cache"))
+				})
 			})
 
-			it("escapes buildpack IDs when restoring buildpack layers", func() {
-				h.AssertNil(t, restorer.Restore(testCache))
-				expectedMetadata := `[metadata]
+			when("there is a cache=true escaped layer", func() {
+				var meta string
+				it.Before(func() {
+					meta = `cache=true
+[metadata]
   escaped-bp-key = "escaped-bp-val"`
-				if txt, err := ioutil.ReadFile(filepath.Join(layersDir, "escaped_buildpack_id", "escaped-bp-layer.toml")); err != nil {
-					t.Fatalf("failed to read escaped-bp-layer.toml: %s", err)
-				} else if !strings.Contains(string(txt), expectedMetadata) {
-					t.Fatalf(`Error: expected '%s' to contain '%s'`, txt, expectedMetadata)
-				}
-
-				expectedText := "echo text from escaped bp layer"
-				if txt, err := ioutil.ReadFile(filepath.Join(layersDir, "escaped_buildpack_id", "escaped-bp-layer", "file-from-escaped-bp")); err != nil {
-					t.Fatalf("failed to read file-from-escaped-bp: %s", err)
-				} else if !strings.Contains(string(txt), expectedText) {
-					t.Fatalf(`Error: expected '%s' to contain '%s'`, txt, expectedText)
-				}
+					h.AssertNil(t, writeLayer(layersDir, "escaped_buildpack_id", "escaped-bp-layer", meta, escapedLayerSHA))
+					h.AssertNil(t, restorer.Restore(testCache))
+				})
+
+				it("keeps layer metadatata", func() {
+					got := h.MustReadFile(t, filepath.Join(layersDir, "escaped_buildpack_id", "escaped-bp-layer.toml"))
+					h.AssertEq(t, string(got), meta)
+				})
+				it("keeps layer sha", func() {
+					got := h.MustReadFile(t, filepath.Join(layersDir, "escaped_buildpack_id", "escaped-bp-layer.sha"))
+					h.AssertEq(t, string(got), escapedLayerSHA)
+				})
+				it("restores data", func() {
+					got := h.MustReadFile(t, filepath.Join(layersDir, "escaped_buildpack_id", "escaped-bp-layer", "file-from-escaped-bp"))
+					want := "echo text from escaped bp layer\n"
+					h.AssertEq(t, string(got), want)
+				})
+			})
+
+			when("there is a cache=true an unknown buildpack", func() {
```

Comment:
This one seems a bit strange to me. I don't think we ever want only metadata for a `cache=true`, `launch=false`, `build=false` layer. It would be hard to get into this situation but in general I think the restorer should never leave the metadata and sha for a cached layer, if it can't restore the layer.

Reply:
Given that it would be nearly impossible to get into this situation, I'm not sure we need to change the implementation, but writing a test around it makes it seem like intentional behavior.

Reply:
The issue here is that the buildpack is not part of the group which makes it difficult to find it. I agree it should be removed. Is there a way to find layers that don't match the group? Should we add it or is it excessive?

Removed the test for now.

Reply:
We could look for unrecognized buildpacks in the layers directory, but given that the analyzer uses the group to create all those directories in the first place, I think it might be excessive.